### DISCUSSION TITLE
Future Plans.

### DISCUSSION THREAD

**Discussion by AethanFoot** [leftwm/leftwm#877] [General]
This is just a simple list of future ideas for leftwm (In no particular order):
- [ ] Optimise calls to xlib/minimise calls to xlib. (Convert to xcb with new release of rust library soon)
- [ ] #558 
- [x] Use the lefthk-core library for keybind handling, and make it feature gated (to allow users to disable built in hotkey daemon and use their own).
- [ ] Change config language (#496). (looking like ron https://github.com/ron-rs/ron)
- [ ] Build up greater configuration and theme settings.
- [ ] Stabilise leftwm-core as a library.
- [ ] Document using leftwm-core as a library.
- [ ] Make leftwm-core display server agnostic (before this we would want the x11 side to be as stable as possible).
- [ ] Add wayland support to leftwm-core then to leftwm. 
- [ ] Add dbus support for ipc (inter-process communication) with [dbus-rs](https://crates.io/crates/dbus). This is for leftwm-state, commands.pipe (maybe keep the file and depreciate that section of code and the examples of using the file. Let leftwm-command handle it) , etc.
- [ ] Add a binary (or expand leftwm-command?) that uses the dbus ipc to get and/or change info about leftwm.

**Comment by grtcdr** [comment]
Hi! Lovely project and great work! 

I'm honestly really psyched to see LeftWM on Wayland!

Quick question: are there any plans to fully comply with the EMWH spec so that `wmctrl` can detect the window manager?

EDIT: seems like it is already EMWH compliant, any particular reason why `wmctrl` just can't detect it?

**Comment by AethanFoot** [comment]
I have added a fix for _NET_WM_NAME and added WM_CLASS to #587. I am assuming you were only talking about 
```wmctrl -m```.

We are always improving EMWH compliance we most updates, it should be at a minimum level (If not it is generally due to a mistake).

**Comment by grtcdr** [comment]
Thank you so much :)

And thank you for the wonderful work you're putting into LeftWM, I really hope it pays off amazingly :heart:

**Comment by ghost** [comment]
This may be a shot to the air but, how about "shorthand" commands in `leftwm-state`? This could be very useful for scripts as a whole.

The best example I have is the bspwm client where:

- `bspc query -D --names` throws the desktop names
- `bspc query -D --names -d focused` throws the focused desktop name

I know `leftwm-state` can use liquid templates but the example in the themes folders seems a bit verbose, it would be nice to have for example: `leftwm-state -w0 --tagnames`. 

I say this might be "shot to the air" because I don't know how much users use lemonbar / xmobar as their actual bars but using an [external program](https://github.com/VentGrey/Epitaph/blob/master/camel#L6) seems / feels odd :sweat_smile: .

Any thoughts on this? :thinking:

**Comment by VuiMuich** [comment]
> I say this might be "shot to the air" because I don't know how much users use lemonbar / xmobar as their actual bars but using an [external program](https://github.com/VentGrey/Epitaph/blob/master/camel#L6) seems / feels odd sweat_smile .

Well, this is basically the Unix way, ins't it?

But in general I think, this would make a great additional "external" sub-command to `leftwm` similar to `leftwm-theme`, so it can be added on users desire. Does anybody know how `cargo` is adding externally installed subcommands without hardcoding them in the main binary?

**Comment by ghost** [comment]
idk if cargo features might help or if a new binary is the solution.

**Comment by VuiMuich** [comment]
I'll investigate, but I thought cargo is able to "add sub commands from binaries at runtime" or even sub commands to sub commands.

**Comment by 0323pin** [comment]
@AethanFoot This,

> Add dbus support for ipc (inter-process communication) with zbus. This is for leftwm-state, commands.pipe (maybe keep the file and depreciate that section of code and the examples of using the file. Let leftwm-command handle it) , etc.

Is it possible to achieve this using `notify-rust` with `default-features = false`, `features = ["d"]` instead?

`zbus` doesn't currently build on NetBSD while `notify-rust` does. Besides, the number of required dependency crates is a lot smaller using `notify-rust`.

Here's the error while building `zbus` on NetBSD

```
   Compiling zbus v1.9.1
error[E0432]: unresolved import `nix::sys::socket::sockopt::PeerCredentials`
   --> /usr/pkgsrc/wip/citron/work/vendor/zbus-1.9.1/src/connection.rs:163:44
    |
163 |         use nix::sys::socket::{getsockopt, sockopt::PeerCredentials};
    |                                            ^^^^^^^^^^^^^^^^^^^^^^^^ no `PeerCredentials` in `sys::socket::sockopt`

error[E0432]: unresolved import `nix::sys::socket::sockopt::PeerCredentials`
   --> /usr/pkgsrc/wip/citron/work/vendor/zbus-1.9.1/src/azync/connection.rs:180:44
    |
180 |         use nix::sys::socket::{getsockopt, sockopt::PeerCredentials};
    |                                            ^^^^^^^^^^^^^^^^^^^^^^^^ no `PeerCredentials` in `sys::socket::sockopt`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `zbus` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

Looking forward to the release of 0.2.10 and to merge the package onto pkgsrc.

**Comment by AethanFoot** [comment]
rust-notify seems only for notifications, but we can use the dbus bindings instead.

**Comment by 0323pin** [comment]
@AethanFoot That should also do it, thanks! I really should look at why `zbus` fails but, if you are ok with dbus bindings, it's fine.

**Comment by mautamu** [comment]
I was running into similar problems on Linux; even once I got `dbus` to compile notifications weren't working for me.

**Comment by KYTAZ0** [comment]
I'd kill for a wayland port as well...

**Comment by Hammer2900** [comment]
How about connecting scripts to python for configuration or event tracking ? We can use this library.
https://github.com/RustPython/RustPython

**Comment by oredaze** [comment]
I didn't want to start a new topic, although I should, because this is a big deal for me: Directional window navigation. Up, down, left, right. When I go to try out a new window manager that is the first thing I look for. Can I focus/swap windows by using the arrows or hjkl? If I can't - it is not worth even trying the window manager. It is just painful for me to not be able to press **Left** and visually focus the window positioned to the left for example... And Leftwm cannot go left right now. (At least I didn't see the functionality)

**Comment by VuiMuich** [comment]
> I didn't want to start a new topic, although I should, because this is a big deal for me: Directional window navigation. Up, down, left, right. When I go to try out a new window manager that is the first thing I look for. Can I focus/swap windows by using the arrows or hjkl? If I can't - it is not worth even trying the window manager. It is just painful for me to not be able to press **Left** and visually focus the window positioned to the left for example... And Leftwm cannot go left right now. (At least I didn't see the functionality)

Hi,
This regularly comes up and I am sorry that this is a dealbreaker for you and some other people as well.
It is not as easy to implement as it might seem on a first glance and most users are very comfortable with the way it is right now.
Of course when you are very used to it, it takes some tedious time to adopt. 

I'll open its own issue though, to document some ideas and thoughts on this topic.
=> #771
